User Inputs

NotJustAVibeCoder opened the pull request:
Added user inputs + validation. For review. Also added comments where (I think) the other functions will go.

NotJustAVibeCoder committed:
User Inputs 

Added user inputs + validation. For review. Also added comments where (I think) the other functions will go.

eperegrine commented on street_level_crime/input_helpers.py:
```diff
@@ -1,14 +1,66 @@
+def validate_postcode(postcode):
+    return bool(uk_postcode_format.match(postcode))
+
+def validate_file_name(file_name):
```
Would it be better here to check if them file actually exists?

```python
import os.path

if os.path.isfile('filename.txt'):
    print ("File exist")
else:
    print ("File not exist")
```
source: https://linuxize.com/post/python-check-if-file-exists/

eperegrine commented on street_level_crime/input_helpers.py:
```diff
@@ -1,14 +1,66 @@
+def ui_get_directory_name():
+    while True:
+        directory_name = input("Please, enter a path to destination folder")
+        if validate_file_name(directory_name):
```
If checking if the actual file exists there would need to be a seperate 'validate_drirectory'

eperegrine commented on street_level_crime/input_helpers.py:
```diff
@@ -1,14 +1,66 @@
+            return file_name
+            break
+        else:
+            print("Wrong! Path cant contain the following characters: < (less than) , > (greater than) , : (colon) , \" (double quote) , / (forward slash)  , \\ (backslash) , | (vertical bar or pipe) \n ? (question mark) , * (asterisk)")
```
File paths should be able to contain slashes?

e.g. `C:\This\Is\A\Path`
or 
`/usr/var/bin/this/is/a/path`

eperegrine commented on street_level_crime/input_helpers.py:
```diff
@@ -1,14 +1,66 @@
+        if validate_file_name(file_name):
+            print("File name accepted")
+            return file_name
+            break
```
```suggestion

```
break not needed after a return

eperegrine commented on street_level_crime/input_helpers.py:
```diff
@@ -1,14 +1,66 @@
+        if validate_file_name(directory_name):
+            print("Path accepted")
+            return directory_name
+            break
```
```suggestion

```

eperegrine commented on street_level_crime/input_helpers.py:
```diff
@@ -1,14 +1,66 @@
+def ui_get_distance():
+    while True:
+        distance = input("Please, select distance from postcode centre: [1] 1km \n [2] 2km \n [3] 5km \n")
+        if distance == str(1) or option == str(2) or option == str(3):
```
I like this, by comparing to the string version of the number you don't have to worry about a type cast error. I've just rewritten it to be a little shorter, we can write the string literal instead of doing a type conversion as we know what the result will be. By having an array it's easier to add more options and I think it's a bit shorter
```suggestion
        if distance in ["1", "2", "3"]:
```

NotJustAVibeCoder committed:
Update street_level_crime/input_helpers.py

Co-Authored-By: Emily Peregrine <8850421+eperegrine@users.noreply.github.com>

NotJustAVibeCoder committed:
Update street_level_crime/input_helpers.py

Co-Authored-By: Emily Peregrine <8850421+eperegrine@users.noreply.github.com>

NotJustAVibeCoder committed:
Update street_level_crime/input_helpers.py

Co-Authored-By: Emily Peregrine <8850421+eperegrine@users.noreply.github.com>

NotJustAVibeCoder committed:
Edits

1. Corrected file validation and made code to check if the file exists in a directory  2. Put RegEx into the functions they're referenced in 3. Combined get_folder_path and get_file_name (don't think we need to collect them separately?) to return full path to a file + '.csv' extension 4. Added folder path formatting (so when user enters C:\path\to\file it changes it to C:\path\to\file\)

eperegrine commented on street_level_crime/input_helpers.py:
```diff
@@ -1,14 +1,84 @@
+    if os.path.isdir(directory):
+        return True
+    else:
+        return False
```
isdir returns a bool so we don't need the if statement
```suggestion
    return os.path.isdir(directory)
```

eperegrine commented on street_level_crime/input_helpers.py:
```diff
@@ -1,14 +1,84 @@
+    if os.path.isfile(path_to_file):
+        return True
+    else:
+        return False
```
```suggestion
    return os.path.isfile(path_to_file)
```

NotJustAVibeCoder replied:
Of course not! I'm new to Python, so it's always good to learn from more experienced coders. Thanks!

NotJustAVibeCoder committed:
Update street_level_crime/input_helpers.py

Co-Authored-By: Emily Peregrine <8850421+eperegrine@users.noreply.github.com>

eperegrine commented on street_level_crime/input_helpers.py:
```diff
@@ -1,14 +1,84 @@
+            print("File name accepted")
+            return directory_formatting(path_to_directory) + file_name + ".csv"
+        
+        elif file_exists(path_to_directory + file_name):
```
The os module has a [path.join method ](https://docs.python.org/3/library/os.path.html#os.path.join) which should remove the need for the directory formatting method

I like that you are allowing the user to not specify extension, the only thing I added was an extra check after adding the extansion to ensure that also exists

```suggestion
        file_name = input("Please enter a file name: ")
        path_to_file=os.path.join(path_to_directory, file_name)
        if not file_exists(path_to_file) and validate_file_name(file_name):
            path_to_file_ext = path_to_file + ".csv"
            if file_exists(path_to_file_ext):
                print("File name accepted")
                return path_to_file + ".csv"
        elif file_exists(path_to_file):
```

eperegrine reviewed (commented):
I hope I don't seem to picky! 😂

I really liked how you you allowed the user to enter a file without the filename

There's no need to put the regular expressions in the function scope, as they need to be compiled it may be better to have them in the global scope as then it only compiles them once, but this isn't a big issue

A good rule of thumb is is if you see
```python
if condition:
    return True
else:
    return False
``` 
you can rewrite it as `return condition`

You could argue that the `directory_exists` and `file_exists` aren't need as they are just wrapping os.path.isfile and isdirectory, but this isn't a huge problem

NotJustAVibeCoder committed:
Update street_level_crime/input_helpers.py

Co-Authored-By: Emily Peregrine <8850421+eperegrine@users.noreply.github.com>

NotJustAVibeCoder committed:
Update street_level_crime/input_helpers.py

Co-Authored-By: Emily Peregrine <8850421+eperegrine@users.noreply.github.com>

NotJustAVibeCoder committed:
Minor changes

Removed break where not required + fixed loop which returns 'path_to_file_ext' + removed directory_formatting function as it's been replaced by os.path.join

NotJustAVibeCoder committed:
Minor changes #2

removed file_exists and directory_exists function as there's no point to have a function for 1 line of code. Also made all functions to return a value.

NotJustAVibeCoder committed:
Regex to compile outside functions
